src/merkle/merkle-objects.ts

```diff
@@ -0,0 +1,94 @@
+import CID from "cids";
+
+import { Doctype } from '@ceramicnetwork/common';
+import {
+  CompareFunction,
+  MergeFunction, MetadataFunction,
+  Node,
+  TreeMetadata,
+} from './merkle';
+
+import { logger } from '../logger';
+import { IpfsService } from '../services/ipfs-service';
+
+import bloom from 'bloomfilter.js';
+
+const BLOOM_FILTER_TYPE = "jsnpm_bloomfilter.js";
+
+export class Candidate {
+  public readonly cid: CID;
+  public readonly document: Doctype;
+  public readonly reqId: number;
+
+  constructor(cid: CID, reqId?: number, document?: Doctype) {
+    this.cid = cid;
+    this.reqId = reqId;
+    this.document = document
+  }
+
+  get docId(): string {
+    return this.document.id.baseID.toString()
+  }
+
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class IpfsMerge implements MergeFunction<Candidate, TreeMetadata> {
+  private ipfsService: IpfsService;
+
+  constructor(ipfsService: IpfsService) {
+    this.ipfsService = ipfsService;
+  }
+
+  async merge(left: Node<Candidate>, right: Node<Candidate>, metadata: TreeMetadata | null): Promise<Node<Candidate>> {
+    const merged = [left.data.cid, right.data.cid];
+    if (metadata) {
+      const metadataCid = await this.ipfsService.storeRecord(metadata);
+      merged.push(metadataCid)
+    }
+
+    const mergedCid = await this.ipfsService.storeRecord(merged);
+    logger.debug('Merkle node ' + mergedCid + ' created.');
+    return new Node<Candidate>(new Candidate(mergedCid), left, right);
+  }
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class IpfsLeafCompare implements CompareFunction<Candidate> {
+  compare(left: Node<Candidate>, right: Node<Candidate>): number {
+    return left.data.docId.localeCompare(right.data.docId);
+  }
+}
+
+/**
+ * Implements IPFS merge CIDs
+ */
+export class BloomMetadata implements MetadataFunction<Candidate, TreeMetadata> {
+  generateMetadata(leaves: Array<Node<Candidate>>): TreeMetadata {
+    const bloomFilterEntries = new Set<string>()
+    for (const node of leaves) {
+      const doc = node.data.document
+      bloomFilterEntries.add(`docid-${doc.id.baseID.toString()}`)
+      bloomFilterEntries.add(`schema-${doc.metadata.schema?.toString()}`)
+      bloomFilterEntries.add(`family-${doc.metadata.family}`) // todo should we use a special charactor of some kind to distingush the family "undefined" from no family?
+      if (doc.metadata.tags) { // TODO should we add a marker for empty tags?
```

Comment:
While it makes sense to include markers for no family and no schema, I was less sure about something for no tags, especially since there can be multiple but the bloom filter doesn't give any other indication of how many tags there are per document.  So I'm leaning towards leaving tags as-is and just not putting any tags-related entries in for documents with no tags

Reply:
Including `no-schema` and `no-schema` makes no sense to me. Why would you want that?

Reply:
Ok, I see your comment above. To me it still doesn't make sense to include them because no one would want to index for "all documents without a schema/family".

Reply:
Interesting.  I agree no one is likely to index *just* on "all documents with no schema".  I could, however, imagine a use case for something like "index all docs with this controller and this family but no schema", or "index all docs with this controller and this tag but no family".  That said I don't feel super strongly about supporting such a use case, so I'm fine with leaving it out for now until/unless someone asks for it down the road

Reply:
> "index all docs with this controller and this family but no schema"

Adding `no-schema` wouldn't really help with this use case since you can't know if this was added for the doucment which you care about or any other documents.